File: CivitasJava/src/civitas/Jugador.java
package civitas;

import java.util.ArrayList;

public class Jugador implements Comparable<Jugador>{

    
    //ATRIBUTOS DE CLASE
    protected static int CasasMax = 4;
    protected static int CasasPorHotel = 4;
    protected static int HotelesMax = 4;
    protected static float PasoPorSalida = 1000;
    protected static float PrecioLibertad = 200;
    private static float SaldoInicial = 7500;
    
    //ATRIBUTOS DE INSTANCIA
    protected boolean encarcelado;
    private boolean puedeComprar;
    private String nombre;
    private int numCasillaActual;
    private float saldo;
    
    //asociaciones
    private ArrayList<TituloPropiedad> propiedades;
    private Sorpresa salvoconducto;
    
    //paquete
    Jugador(String nombre){
        this.nombre = nombre;
        encarcelado = false;
        puedeComprar = false;
        numCasillaActual = 0;
        saldo = SaldoInicial;
        propiedades = new ArrayList<TituloPropiedad>();
        salvoconducto = null;
    }
    
    protected Jugador(Jugador otro)
    {
        this.encarcelado=otro.encarcelado;
        this.nombre=otro.nombre;
        this.numCasillaActual=otro.numCasillaActual;
        this.puedeComprar=otro.puedeComprar;
        this.saldo=otro.saldo;
        this.propiedades=otro.propiedades;
        this.salvoconducto=otro.salvoconducto;
    }
    
    boolean cancelarHipoteca(int ip)
    {   

        boolean result=false;
        if(!encarcelado)
        {
            if(existeLaPropiedad(ip))
            {
               TituloPropiedad propiedad=propiedades.get(ip);
               float cantidad=propiedad.getImporteCancelarHipoteca();
               boolean puedoGastar=puedoGastar(cantidad);
               if(puedoGastar)
               {
                   result=propiedad.cancelarHipoteca(this);
                   
                   if(result)
                   {
                       Diario.getInstance().ocurreEvento("El jugador"+nombre+"cancela la hipoteca de la propiedad"+ip);
                       
                   }
                   
               }
            }
        }
        
        return result;
    }
    
    
    int cantidadCasasHoteles()
    {
        int resultado=0;
        for(int i=0;i<propiedades.size();i++)
        {
            resultado=propiedades.get(i).cantidadCasasHoteles();
        }
        
        return resultado;
        
         
    }
    
    boolean comprar(TituloPropiedad titulo){
        boolean result=false;
            if(!encarcelado)
            {
                if(puedeComprar)
                {
                    float precio=titulo.getPrecioCompra();
                    if(puedoGastar(precio))
                    {
                        result=titulo.comprar(this);
                        if(result)
                        {
                            propiedades.add(titulo);
                            Diario.getInstance().ocurreEvento("El jugador"+this+ " compra la propiedad "+titulo.toString());
                        }
                        puedeComprar=false;
                    }
                }
            }
        return result;
    }
    
    boolean construirCasa(int ip){
        boolean result=false;
            
            if(encarcelado)
            {
                return result;
            }
            else
            {
                boolean puedoEdificarCasa=false;
                boolean existe=existeLaPropiedad(ip);
                if(existe)
                {
                    TituloPropiedad propiedad=propiedades.get(ip);
                    puedoEdificarCasa=puedoEdificarCasa(propiedad);
                    float precio=propiedad.getPrecioEdificar();
                    if(puedoGastar(precio) && puedoEdificarCasa)
                    {
                        result=propiedad.construirCasa(this);
                        Diario.getInstance().ocurreEvento("El jugador"+nombre+" construye casa en la propiedad"+ip);
                    }
                }
            }
            
        return result;
    }
    
    boolean construirHotel(int ip){
        
        boolean result=false;
        if(!encarcelado)
        {
            if(existeLaPropiedad(ip))
            {
                TituloPropiedad propiedad=propiedades.get(ip);
                boolean puedoEdificarHotel=puedoEdificarHotel(propiedad);
                puedoEdificarHotel=false;
                float precio=propiedad.getPrecioEdificar();
                
                if(puedoGastar(precio) && puedoEdificarHotel && propiedad.getNumCasas()==getCasasPorHotel())
                {
                       puedoEdificarHotel=true;
                }
                
                if(puedoEdificarHotel)
                {
                    result=propiedad.construirHotel(this);
                    int casasPorHotel=getCasasPorHotel();
                    propiedad.derruirCasas(casasPorHotel,this);
                    Diario.getInstance().ocurreEvento("El jugador "+nombre+ " construye hotel en la propiedad "+ip);
                }
            }
        }
        
        return result;
    }
    
    Boolean enBancarrota()
    {
        Boolean resultado=false;
        if(saldo<0)
        {
            resultado=true;
        }
        return resultado;
    }
    
    boolean encarcelar(int numCasillaCarcel){
        
        if(debeSerEncarcelado()){
            
            moverACasilla(numCasillaCarcel);
            encarcelado = true;
            Diario.getInstance().ocurreEvento("Se ha encarcelado al jugador "+nombre);
            
        }
        
        return encarcelado;
    }
    
    int getCasasPorHotel(){
        return CasasPorHotel;
    }
    
    int getNumCasillaActual(){
        return numCasillaActual;
    }
    
    String getNombre(){
        return nombre;
    }
    
    boolean getPuedeComprar(){
        return puedeComprar;
    }
    
    boolean hipotecar(int ip){
        boolean result=false;
        if(!encarcelado)
        {
            if(existeLaPropiedad(ip))
            {
                TituloPropiedad propiedad=propiedades.get(ip);
                result=propiedad.hipotecar(this);

            }
            if(result)
            {
                Diario.getInstance().ocurreEvento("El jugador "+nombre+ " hipoteca la propiedad "+ip);
            }
        }
        return result;
    }
    
    boolean modificarSaldo(float cantidad){
        
        saldo += cantidad;
        Diario.getInstance().ocurreEvento("Saldo de "+nombre+" incrementado(+"+cantidad+")");
        return true;
    }
    
    boolean moverACasilla(int numCasilla){
        boolean resultado = false;
        if(!encarcelado){
            numCasillaActual = numCasilla;
            puedeComprar = false;
            Diario.getInstance().ocurreEvento(nombre+" se ha movido a la casilla "+numCasillaActual);
            resultado = true;
        }
        
        return resultado;
    }
    
    boolean obtenerSalvoConducto(Sorpresa sorpresa){
        boolean seObtieneSalvoConducto = encarcelado;
        
        if(!seObtieneSalvoConducto){
            salvoconducto = sorpresa;
            
        }
        
        return seObtieneSalvoConducto;
        
    }
    
    boolean paga(float cantidad){
        
        boolean resultado = modificarSaldo(cantidad*-1);
        
        return resultado;
    }
    
    boolean pagaAlquiler(float cantidad){
        boolean resultado = false;
        
        if(!encarcelado)
            resultado = paga(cantidad);
        
        return resultado;
    }
    
    boolean pagaImpuesto(float cantidad){
        boolean resultado = false;
        
        if(!encarcelado)
            resultado = paga(cantidad);
        
        return resultado;
        
        
    }
    
    boolean pasaPorSalida(){
        
        modificarSaldo(PasoPorSalida);
        Diario.getInstance().ocurreEvento(nombre+" ha pasado por la salida");
        return true;
    }
    
            
    boolean puedeComprarCasilla(){
       
        puedeComprar = !encarcelado;
        
        return puedeComprar;
    }   
    
    boolean recibe(float cantidad){
        
        boolean resultado = false;
        
        if(!encarcelado)
            resultado = modificarSaldo(cantidad);
        
        return resultado;
        
    }
    
    boolean salirCarcelPagando(){
        boolean resultado = false;
        
        if(puedeSalirCarcelPagando()){
            paga(PrecioLibertad);
            encarcelado=false;
            Diario.getInstance().ocurreEvento(nombre+" ha sido puesto en libertad");
            resultado = true;
        }
        
        return resultado;
    }
    
    boolean salirCarcelTirando(){
        
        boolean resultado = Dado.getInstance().salgoDeLaCarcel();
        
        if(resultado){
            encarcelado = false;
            Diario.getInstance().ocurreEvento(nombre+" ha salido de la cárcel");
            resultado = true;
        }
        
        return resultado;
    }
    
    boolean tieneAlgoQueGestionar()
    {    
        return !propiedades.isEmpty();
    }
    
    Boolean tieneSalvoconducto()
    {
        Boolean resultado=false;
        if(salvoconducto!=null)
        {
            resultado=true;
        }
        
        return resultado;
    }
    
    boolean vender(int ip){
        
        boolean resultado = false;
        
        if(!encarcelado){
            
            if(existeLaPropiedad(ip)){
               resultado = propiedades.get(ip).vender(this);
            }
            
            if(resultado){
                Diario.getInstance().ocurreEvento(nombre+" vende la propiedad "+propiedades.get(ip).getNombre());
                
                propiedades.remove(ip);
            }
            
        }
        
        return resultado;
    }
    
    //publicos
    @Override
    public int compareTo(Jugador otro){

        /*
        if a < b then return -1
            if a = b then return  0
            if a > b then return  1
            if a and b no son comparables return null

        */
        int resultado = Float.compare(saldo, otro.saldo);

        return resultado;
    }
    
    public boolean isEncarcelado(){
        
        return encarcelado;
    }
    
    @Override
    public String toString(){
                
        String datos="Nombre:"+nombre+"\nSaldo: "+saldo+"\nEncarcelado:"+encarcelado+
                "\nCasilla actual:"+numCasillaActual+
                "\nPuede comprar: "+puedeComprar+
                "\nSalvoconducto: "+tieneSalvoconducto();
        
        return datos;
    }
    
    
    //privados
    
    Boolean existeLaPropiedad(int ip)
    {
        Boolean resultado=false;
        if(!propiedades.isEmpty())
        {
          if(propiedades.get(ip).getPropietario()==this)
            {
                resultado=true;
            }
        }

        return resultado;
    }
    
    
 private int getCasasMax()
    {
        return CasasMax;
    }
    
    int getHotelesMax()
    {
        return HotelesMax;
    }

    
    private float getPrecioLibertad()
    {
        return PrecioLibertad;
    }
    
    private float getPremioPasoSalida()
    {
       return PasoPorSalida; 
    }
    
    public ArrayList<TituloPropiedad> getPropiedades()
    {
        return propiedades;
    }
    
    public int getSizePropiedades()
    {
        return propiedades.size();
    }
    
    private void perderSalvoConducto(){
        
        salvoconducto.usada();
        salvoconducto = null;
        
    }
    
    private boolean puedeSalirCarcelPagando(){
        return (saldo >= PrecioLibertad);  
    }
    
    private Boolean puedoEdificarCasa(TituloPropiedad propiedad)
    {
        Boolean resultado=false;
        if(propiedad.getNumCasas()<CasasMax)
        {
            resultado=true;
        }
        
        return resultado;       
        
    }
    
    private Boolean puedoEdificarHotel(TituloPropiedad propiedad)
    {
        Boolean resultado=false;
        if(propiedad.getNumHoteles()<HotelesMax)
        {
            resultado=true;
        }
        
        return resultado;    
    }
    
    private boolean puedoGastar(float precio){
        
        boolean resultado = false;
        
        if(!encarcelado){
            resultado = (saldo>=precio);
        }
        
        return resultado;
    }
        
    //protected    
        protected float getSaldo()
    {
        return saldo;
    }
    
    protected boolean debeSerEncarcelado(){
        
        boolean seEncarcela = false;
        
        if(!encarcelado){
           if(tieneSalvoconducto()){
               perderSalvoConducto();
               Diario.getInstance().ocurreEvento("El jugador "+nombre+" se ha librado de la cárcel.");
           }else
               seEncarcela = true;
        }
            
        return seEncarcela;
    }
    public static float getSaldoInicial() {
        return SaldoInicial;
    }


    public Sorpresa getSalvoconducto() {
        return salvoconducto;
    }
    
    public static void main(String [] args){
        
        Jugador j1=new Jugador("Roberto");
        Jugador j2=new Jugador("Andres");
                j1.compareTo(j2);
        
        System.out.print(j1.toString());
        
        //j1.cantidadCasasHoteles();
        j1.debeSerEncarcelado();
        j1.enBancarrota();
        //j1.existeLaPropiedad(2);
        j1.getSaldo();
        j1.recibe(SaldoInicial);
        j1.puedoGastar(328943953);
    }

}

File: CivitasRuby/lib/jugador.rb
#encoding:utf-8


require_relative 'diario'
require_relative 'sorpresa'
require_relative 'mazo_sorpresas'
require_relative 'tipo_sorpresa'
require_relative 'titulo_propiedad'
require_relative 'jugador'


module Civitas
 
    class Jugador


      @@casas_max=4
      @@casas_por_hotel=4
      @@hoteles_max=4
      @@paso_por_salida=1000.0
      @@precio_libertad=200.0

      attr_reader :nombre , :num_casilla_actual , :encarcelado , :casas_por_hotel , :puede_comprar,
        :hoteles_max , :precio_libertad , :paso_por_salida , :propiedades , :saldo , :salvoconducto 
      
      attr_accessor :nombre , :num_casilla_actual , :encarcelado , :casas_por_hotel , :puede_comprar,
        :hoteles_max , :precio_libertad , :paso_por_salida , :propiedades , :saldo , :salvoconducto 

      private :precio_libertad ,:paso_por_salida


      def initialize(encarcelado,nombre,num_casilla_actual,puede_comprar,saldo,salvoconducto,propiedades)
        @encarcelado = encarcelado
        @nombre = nombre
        @num_casilla_actual = num_casilla_actual
        @puede_comprar = puede_comprar
        @saldo = saldo
        @salvoconducto = salvoconducto
        @propiedades = propiedades
      end

      def self.new_por_nombre(nombre)

        new(false,nombre,0,false,95000,nil,Array.new)

      end
      

      def self.new_por_copia(jugador)
        new(jugador.encarcelado, jugador.nombre, jugador.num_casilla_actual, jugador.puede_comprar, jugador.saldo,
          jugador.salvoconducto, jugador.propiedades)
      end
      
      def compare_to(otro)
=begin
a <=> b :=
  if a < b then return -1
  if a = b then return  0
  if a > b then return  1
  if a and b are not comparable then return nil
=end
      @saldo<=>otro.saldo
        
      end

      def is_encarcelado

      end

      def to_string   

        datos = "Nombre del jugador: #{@nombre} Saldo del jugador: #{@saldo}
Número de propiedades del jugador: #{@propiedades.length()}
Número casilla actual del juagdor: #{@num_casilla_actual}"

        datos;
      end

      def comprar(titulo)
        
        result = false
        
        if(@encarcelado)
          return result
        
        elsif(@puede_comprar)
          
          precio = titulo.precio_compra
          
          if(puedo_gastar(precio))
            result = titulo.comprar(self)
          
          
            if(result)
              @propiedades<<titulo
              Diario.instance.ocurre_evento("++#{@nombre} ha adquirido la propiedad #{titulo.nombre}++")
              @puede_comprar = false
            end
          end
        end
      result
      end

      def cancelar_hipoteca(ip)
        result = false; 
        
        if(@encarcelado)
          return result
        
        elsif(existe_la_propiedad(ip) and @propiedades[ip].hipotecado)
          propiedad = @propiedades[ip];
          cantidad = propiedad.get_importe_cancelar_hipoteca
          
          if(puedo_gastar(cantidad))
            result = propiedad.cancelar_hipoteca(self)
            
            if(result)
              Diario.instance.ocurre_evento("#{@nombre} cancela la hipoteca de la propiedad #{ip}")
            end
            
          end
          result
        end
      end

      def cantidad_casas_hoteles
        suma = 0
        for i in(0..@propiedades.length)
          suma += @propiedades[i].num_casas
          suma += @propiedades[i].num_hoteles
        end
        
        suma
      end



      def construir_casa(ip)
        result=false
        
        if(@encarcelado)
          return result
        else
          existe = existe_la_propiedad(ip)
          if(existe)
            
            propiedad = @propiedades[ip]
            puedo_edificar = puedo_edificar_casa(propiedad)
            precio = propiedad.precio_edificar
            
            if(puedo_gastar(precio) and propiedad.num_casas < @@casas_max)
              puedo_edificar = true
            end
            
            if(puedo_edificar)
              result = propiedad.construir_casa(self)
            end
              
            if(result)
              Diario.instance.ocurre_evento("#{@nombre} construye una casa en #{@propiedades[ip].nombre}")
            end
          end
        end
        result
      end

      def construir_hotel(ip)
        result = false
        
        if(@encarcelado)
          return result
        
        elsif(@existe_la_propiedad[ip])
          propiedad = @propiedades[ip]
          puedo_edificar = puedo_edificar_hotel(propiedad)
          precio = propiedad.precio_edificar
          
          if(puedo_gastar(precio) and propiedad.num_hoteles < @@hoteles_max and propiedad.num_casas == @@casas_max)
            puedo_edificar = true
          end
          
          if(puedo_edificar)
            result = propiedad.construir_hotel(self)
            propiedad.derruir_casas(@@casas_por_hotel,self)
            Diario.instance.ocurre_evento("#{@nombre} construye un hotel en #{@propiedades[ip].nombre}")
          end
        end
        result
      end


      def en_bancarrota
        devolucion=false
        
        if(@saldo<= 0 )
            devolucion=true
        end

        devolucion
      end

      def encarcelar(num_casilla_carcel)
        
          if debe_ser_encarcelado
            
            if(mover_a_casilla(num_casilla_carcel))
              @encarcelado = true
              Diario.instance.ocurre_evento("#{@nombre} entra en la prisión")
            else
              Diario.instance.ocurre_evento{"#{@nombre} no ha podido entrar en Prisión"}
            end
          end
          
          @encarcelado
      end

      def debe_ser_encarcelado

        if @encarcelado
          devolver = false
        
        elsif tiene_salvoconducto
          perder_salvoconducto
          devolver = false
          Diario.instance.ocurre_evento("#{@nombre} utiliza la carta SalvoConducto y se libra de Prisión")
        
        else
          devolver = true
        end
        devolver
      end

      

      def hipotecar(ip)
        result=false
        
        if(@encarcelado)
          return result
        elsif(existe_la_propiedad(ip))
          propiedad = @propiedades[ip]
          result = propiedad.hipotecar(self)
        end
        
        if(result)
          Diario.instance.ocurre_evento("El jugador #{@nombre} hipoteca su propiedad #{@propiedades[ip].nombre}")
        end
        
        result

      end


      def modificar_saldo(cantidad)
        
        @saldo+=cantidad
        Diario.instance.ocurre_evento("Se ha modificado el saldo de #{@nombre}(#{cantidad})")
        return true
      end

      def mover_a_casilla(num_casilla)
        
        if(@encarcelado)
          devolver = false
        else
          Diario.instance.ocurre_evento("#{@nombre} se ha movido de #{@num_casilla_actual} a #{num_casilla}")
          @num_casilla_actual = num_casilla
          @puede_comprar = false
          devolver = true
        end
        
        devolver
      end

      def obtener_salvoconducto(sorpresa)

        if (@encarcelado)
          devolver = false
        
        else
        
          @salvoconducto = sorpresa
          devolver = true
        end
        
        devolver
      end

      def paga(cantidad)

        return modificar_saldo(cantidad*-1)
      end

      
      def paga_alquiler(cantidad)
        
        if(@encarcelado)
          devolver = false
        else
        
          devolver = paga(cantidad)
        end
        devolver
      end

      def paga_impuesto(cantidad)

        if(@encarcelado)
          devolver = false
        else 
        
          devolver = paga(cantidad)
        
        end
        return devolver
      end

      def pasa_por_salida()
          modificar_saldo(@@paso_por_salida)
          Diario.instance.ocurre_evento("#{@nombre} ha pasado por la salida")
      end

      def   puede_comprar_casilla
        
        @puede_comprar = true
        
        if(@encarcelado)
          @puede_comprar = false
        end
        
        @puede_comprar
        
      end

      def recibe(cantidad)
      
        if(@encarcelado)
          resultado = false
        else
          resultado = modificar_saldo(cantidad)
        end
        resultado
      end

      def salir_carcel_pagando()
        devolver = false
        
        if(@encarcelado and puede_salir_carcel_pagando)
          
          paga(@@precio_libertad)
          @encarcelado = false
          Diario.instance.ocurre_evento("#{@nombre} ha pagado para salir de la Prisión")
        end
        
        devolver
      end

      def salir_carcel_tirando()
        
        devolver = false
        
        if (Dado.instance.salgoDeLaCarcel)
          @encarcelado = false
          Diario.instance.ocurre_evento("#{@nombre} tira el Dado para salir de la Prisión")
        end
        devolver
      end

      def   tiene_algo_que_gestionar()
      
        devolver = false
        
        if(@propiedades.length > 0 )
          devolver = true
        end
        devolver
      end

      def tiene_salvoconducto()
        devolver = false
        if(@salvoconducto != nil)
          devolver = true
        end
        devolver
      end

      def vender(ip)
        
        if(@encarcelado)
          devolver = false
        else
        
          if(existe_la_propiedad(ip) and !@propiedades[ip].hipotecado)
            propiedad = @propiedades[ip]
            devolver = propiedad.vender(self)
            Diario.instance.ocurre_evento("#{@nombre} ha vendido la propiedad #{@propiedades[ip].nombre}")
            @propiedades.delete_at(ip)
          end
        end
        devolver
      end
      
      private
      def   perder_salvoconducto
          @salvoconducto.usada
          @salvoconducto=nil
      end
      
      def existe_la_propiedad(ip)
        
        devolver = false
        
        for i in(0..@propiedades.length-1)
          if @propiedades[i].nombre == ip
            devolver = true
          end
        end
        devolver
      end
      
      def puede_salir_carcel_pagando
      
        devolver = false
        
        if @saldo >= @@precio_libertad
          devolver = true
        end
        
        devolver
        
      end
      
      def   puedo_edificar_casa(propiedad)
        return propiedad.num_casas<=@@casa_max
      end
      
      def   puedo_edificar_hotel(propiedad)
        return propiedad.num_hoteles<=@@hoteles_max
      end
      
      def puedo_gastar(precio)
        
        if(@encarcelado)
          devolver = false
        
        else
          
          if (@saldo >= precio)
            devolver = true
          end
        end
        devolver
      end
          
     private_class_method :new

    end
end

j = Civitas::Jugador.new_por_nombre("Antonio")